internal/datastore/crdb/watch.go

```diff
@@ -250,200 +68,272 @@ func (cds *crdbDatastore) WatchSchema(ctx context.Context, afterRevision datasto
+		return
+	}
+
+	// Default: 1s
+	resolvedDuration := 1 * time.Second
+	if opts.CheckpointInterval > 0 {
+		resolvedDuration = opts.CheckpointInterval
+	}
+
+	resolvedDurationString := strconv.FormatInt(resolvedDuration.Milliseconds(), 10) + "ms"
+	interpolated := fmt.Sprintf(cds.beginChangefeedQuery, strings.Join(tableNames, ","), afterRevision, resolvedDurationString)
+
+	sendError := func(err error) {
 		if errors.Is(ctx.Err(), context.Canceled) {
-			processUpdate(nil, datastore.NewWatchCanceledErr())
+			errs <- datastore.NewWatchCanceledErr()
 			return
 		}
 
-		processUpdate(nil, err)
+		if pool.IsResettableError(ctx, err) || pool.IsRetryableError(ctx, err) {
+			errs <- datastore.NewWatchTemporaryErr(err)
+			return
+		}
+
+		errs <- err
+	}
+
+	sendChange := func(change *datastore.RevisionChanges) bool {
+		select {
+		case updates <- change:
+			return true
+
+		default:
+			errs <- datastore.NewWatchDisconnectedErr()
+			return false
+		}
+	}
+
+	changes, err := conn.Query(ctx, interpolated)
+	if err != nil {
+		sendError(err)
 		return
 	}
 
 	// We call Close async here because it can be slow and blocks closing the channels. There is
 	// no return value so we're not really losing anything.
 	defer func() { go changes.Close() }()
 
+	tracked := common.NewChanges(revision.DecimalKeyFunc, opts.Content)
+
 	for changes.Next() {
 		var tableNameBytes []byte
 		var changeJSON []byte
 		var primaryKeyValuesJSON []byte
 
+		// Pull in the table name, the primary key(s) and change information.
 		if err := changes.Scan(&tableNameBytes, &primaryKeyValuesJSON, &changeJSON); err != nil {
-			if errors.Is(ctx.Err(), context.Canceled) {
-				processUpdate(nil, datastore.NewWatchCanceledErr())
-				return
-			}
-
-			processUpdate(nil, err)
+			sendError(err)
 			return
 		}
 
-		var details schemaChangeDetails
+		var details changeDetails
 		if err := json.Unmarshal(changeJSON, &details); err != nil {
-			processUpdate(nil, err)
+			sendError(err)
 			return
 		}
 
+		// Resolved indicates that the specified revision is "complete"; no additional updates can come in before or at it.
+		// Therefore, at this point, we issue tracked updates from before that time, and the checkpoint update.
 		if details.Resolved != "" {
-			revision, err := cds.RevisionFromString(details.Resolved)
+			rev, err := cds.RevisionFromString(details.Resolved)
 			if err != nil {
-				processUpdate(nil, fmt.Errorf("malformed resolved timestamp: %w", err))
+				sendError(fmt.Errorf("malformed resolved timestamp: %w", err))
 				return
 			}
 
-			processUpdate(&datastore.SchemaState{
-				Revision:     revision,
-				IsCheckpoint: true,
-			}, nil)
+			for _, revChange := range tracked.FilterAndRemoveRevisionChanges(revision.DecimalKeyLessThanFunc, rev.(revision.Decimal)) {
+				revChange := revChange
+				if !sendChange(&revChange) {
+					return
+				}
+			}
+
+			if opts.Content&datastore.WatchCheckpoints == datastore.WatchCheckpoints {
```

Comment:
It seems like we don't have a test for checkpoints - run all CRDB tests with code coverage

Reply:
We don't because they aren't used externally yet

Reply:
It's still code written and it's at risk of being forgotten down the line.